mesonbuild/environment.py

```diff
@@ -1677,7 +1679,11 @@ def _detect_objc_or_objcpp_compiler(self, for_machine: MachineChoice, objc: bool
         self._handle_exceptions(popen_exceptions, compilers)
 
     def detect_java_compiler(self, for_machine):
-        exelist = self.lookup_binary_entry(for_machine, 'java')
+        java_home = self.properties[for_machine].get_java_home()
+        if java_home is not None:
+            exelist = [str(pathlib.Path(java_home) / 'bin' / 'javac')]
+        else:
+            exelist = self.lookup_binary_entry(for_machine, 'javac')
```

Comment:
You might want to look at the C# detection logic below this, since we now potentially have two places to look ($JAVA_HOME/bin and $PATH), also, the `lookup_binary_entry` should be the first choice, the JAVA_HOME should be the second.

You might need to add an optional parameter to the `Environment._get_compilers` method, or do something clever with the `self.default_java` attribute. I'd probably add an optional keyword argument though. Maybe `prepend_candidates`? This is important because we guaranteed that if you put a compiler in your machine file or use CC=clang that we will either give you clang for your c compiler, or error out.

Reply:
Putting `lookup_binary_entry` first will also fix the test failures I think.

Reply:
Feels weird that we look for the includes in $JAVA_HOME, but look for the binary if it appears on $PATH. Because $JAVA_HOME/bin/javac isn't necessarily javac that is on the $PATH.

I am also wondering if maybe looking at $JAVA_HOME is dumb, and instead we should essentially do a `dirname $(dirname $(readlink -f $(which javac)))` to get to the include directory. This is easily supported using `pathlib`. This has the caveat that assumes javac on $PATH is a symlink, but maybe if it isn't a symlink we can fallback to the current behavior.

Reply:
This has the added benefit of relying on system tooling for toggling Java (alternatives on Fedora, etc.)

Reply:
The lookup_binary_entry doesn't check $PATH, that's checked if someone says `JAVAC=/path/to/some/javac meson builddir` or sets `[binaries]javac = '/some/path'` in their machine file. That needs to be checked first.

After that I'm opened to doing whatever would be best (checking to see if javac is a symlink, looking in $JAVA_HOME/bin, whatever.

Reply:
Oh ok had a misunderstanding. Now I get it. I will fix it appropriately later today. Thanks for your explanations

Reply:
no problem, thanks for working on this!